utils/clflags.ml

```diff
@@ -445,9 +445,11 @@ end
 let stop_after = ref None (* -stop-after *)
 
 let should_stop_after pass =
-  match !stop_after with
-  | None -> false
-  | Some stop -> Compiler_pass.rank stop <= Compiler_pass.rank pass
+  if pass = Compiler_pass.Typing && !print_types then true
+  else
+    match !stop_after with
+    | None -> false
```

Comment:
With this implementation, `-i` has a higher priority than `-stop-after`.
A slightly different implementation would be to reverse the priority and let `-i` stop the compilation after typing only if `stop_after` was not set itself.

Reply:
The way I think of it, ideally `-i` should be decomposable in several simpler options, and in particular it implies `-stop-after typing`. Now if two different `-stop-after <pass>` are passed on the command-line, I would expect to stop at the earliest pass specified. This means that if `-i -stop-after <pass>` is specified, I would expect to stop at the earliest pass of `typing` and `<pass>`. Does that make sense?

Reply:
Currently, it's an error two pass two different -stop-after <pass> options. 
https://github.com/ocaml/ocaml/blob/a5e121621d58eab4d232d0788b13d2576fe53d18/driver/main_args.ml#L1834-L1843

Reply:
Apropos https://github.com/ocaml/ocaml/pull/8938#issuecomment-549081223, remember that `-i` needs to decompose to *three* options (or there needs to be another micro-step before `stop-after typing`) to ensure that `.cmi` files are not written.

`-i` already has very high priority - it overrides both `-c` and `-o` (in the sense that `ocamlopt -o foo foo.ml -i` doesn't produce an executable or any other files).

Reply:
I didn't remember this, thanks! This suggests that we could also disallow passing `-i` and `-stop-after` at the same time. (But then that would suggest making the useful side-effect of `-i` available with for example a `-dtypes` or `-dsignature` option.) But personally I think that the intersection behavior would also be sensible, and better than giving priority to one option or the other.

Meta-level point: I personally don't care very strongly about what `-i` does or does not, it's a fairly minor feature that is probably not worth spending too much time on it. I applaud your (@graty-js) care to leave the details of the command-line options well-rounded, but it would also probably be fine to leave it as is, or even slightly changed.

Reply:
> This suggests that we could also disallow passing `-i` and `-stop-after` at the same time.

I considered it, but -i already silently ignores some other options, such as -c and -S, so it could also silently ignore -stop-after <pass> when pass is after Typing.

Reply:
Yes, I think it would be fine that way.

Reply:
Regarding the meta-level point, I agree with you in principle - the only problem is that there's an expectation that `ocamlc -i foo.ml > foo.mli` should give you a reasonable start for an `.mli` file (I can't remember if it's a valid one - i.e. if there are cases where it's not true). I don't know if any bad build systems have capitalised on that (and so might be broken by `-i` starting to produce other artefacts).

Reply:
ocamlbuild has a `foo.inferred.mli` target that creates the file using this process, but it is only invoked manually as a user-provided target, and in any case I don't think anything would break if `.cmi` files were suddenly produced. (I don't know if any build systems is able to track precisely which files are produced as a byproduct of certain command-line flag choices, but neither ocamlbuild nor make try it.)